File: src/main/java/com/example/accountbook/model/Category.java
package com.example.accountbook.model;

import javax.persistence.*;
import javax.validation.constraints.NotNull;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"name", "main_category_id"}))
public class Category {

    public static final int TYPE_INCOME = 1;
    public static final int TYPE_EXPENSE = 2;
    public static final int TYPE_SAVE = 3;
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(optional = false)
    private MainCategory mainCategory;

    @NotNull
    private String name;

    @NotNull
    private int type;

    public Long getId() {
        return id;
    }

    public MainCategory getMainCategory() {
        return mainCategory;
    }

    public void setMainCategory(MainCategory mainCategory) {
        this.mainCategory = mainCategory;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }
}

File: src/main/java/com/example/accountbook/model/Category.kt
package com.example.accountbook.model

import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["name", "main_category_id"])])
data class Category(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long,
    @ManyToOne(optional = false)
    var mainCategory: MainCategory,
    var name: @NotNull String,
    var type: @NotNull Int = 0
) {
    companion object {
        const val TYPE_INCOME = 1
        const val TYPE_EXPENSE = 2
        const val TYPE_SAVE = 3
    }
}
